notebooks/automl.clj

```diff
@@ -0,0 +1,218 @@
+(->>
+ (ml/train train-ds {:model-type :metamorph.ml/dummy-classifier})
+ (ml/predict test-ds)
+ :survived)
+
+;;  the same with pipelines
+(def pipeline
+  (mm/pipeline (ml/model {:model-type :metamorph.ml/dummy-classifier})))
+(->>
+ (mm/fit-pipe train-ds pipeline)
+ (mm/transform-pipe test-ds pipeline)
+ :metamorph/data :survived)
+
+
+;;  but know we can add more operations to the pipeline,
+;;  and nothing else changes, for example drop columns.
+(def pipeline-2
+  (mm/pipeline
+   (mm/lift tc/drop-columns [:embarked])
+   (ml/model {:model-type :metamorph.ml/dummy-classifier})))
+(->>
+ (mm/fit-pipe train-ds pipeline-2)
+ (mm/transform-pipe test-ds pipeline-2)
+ :metamorph/data :survived)
+
+;; The 'lift` function transposes a dataset->dataset function into a ctx->ctx function, as required for
+;; metamorph pipeline operations
+;;
+;; So a metamorph pipeline can encapsulate arbitray transformation
+;; of a dataset in the 2 modes. They can be "stateless"
+;; (only chaining the dataset, such as `drop-columns`) or
+;; "state-full", so they store data in the ctx during :fit and can use
+;; it in :transform. In the pipeline above, the trained model is
+;; stored in this way.
+;;
+;;This makes pipeline execution "isolated"
+;;
+;; For convenience `tablecloth` contains a ns where all functions
+;; are lifted into ctx->ctx operations, so can be added to pipelines
+;; directly.
+(require 'tablecloth.pipeline)
+(def pipeline-3
+  (mm/pipeline
+   (tablecloth.pipeline/drop-columns [:embarked])
+   (ml/model {:model-type :metamorph.ml/dummy-classifier})))
+(->>
+ (mm/fit-pipe train-ds pipeline-3)
+ (mm/transform-pipe test-ds pipeline-3)
+ :metamorph/data :survived)
+
+;;  It supports as well "pipelines as data"
+(def pipeline-4
+  (mm/->pipeline
+   [[tablecloth.pipeline/drop-columns [:embarked]]]
+   [[ml/model {:model-type :metamorph.ml/dummy-classifier}]]))
+(->>
+ (mm/fit-pipe train-ds pipeline-4)
+ (mm/transform-pipe test-ds pipeline-4)
+ :metamorph/data :survived)
+
+;; The auto ml support in metamorph consists now in the possibility
+;; to create an arbitrary number of different pipelines
+;; and combine them with arbitray test/train data splits
+
+;;  helper for later
+(defn make-results-ds [evaluation-results]
+  (->> evaluation-results
+       flatten
+       (map #(hash-map :options (-> % :test-transform :ctx :model :options)
+                       :used-features (-> % :fit-ctx :used-features)
+                       :mean-accuracy (-> % :test-transform :mean)))
+       tc/dataset))
+
+
+(require '[scicloj.metamorph.ml :as ml]
+         '[scicloj.metamorph.ml.loss :as loss]
+         '[scicloj.metamorph.core :as mm]
+
+         '[tablecloth.api :as tc])
+
+;; ## Finding best model automatcally
+;;  This advantage of the pipelines is even more visible,
+;;  if we want to have configurable pipelines,
+;;  and do a grid search to find optimal settings
+
+;;  the following will find teh best model across:
+;;  5 different model classes
```

Comment:
Note that splitting Markdown into separate lines does not result in separate lines in the resulting HTML.